server/aesdsocket.c

```diff
@@ -218,124 +343,238 @@ void socket_connect()
+	if (output_buffer == NULL)
+	{
+		printf("Malloc failed!\n");
+		exit(1);
+	}
+	memset(output_buffer, 0, BUFFER_SIZE);
+	// For test
+
+	/*Packet reception, detection and storage logic*/
+	while (packet_comp == false)
+	{
+
+		// printf("Receiving data from descriptor:%d.\n",sfd);
+
+		ret_recv = recv(params->client_fd, buff, BUFFER_SIZE, 0); //**!check the flag
+		if (ret_recv < 0)
 		{
-			printf("Error: Opening file\n");
-			syslog(LOG_ERR, "Error: Opening file %s. Exiting ", strerror(errno));
+			printf("Error while receving data packets\n");
+			syslog(LOG_ERR, "Error: Receiving failed =%s. Exiting ", strerror(errno));
 			exit(EXIT_FAILURE);
 		}
+		else if (ret_recv == 0)
+		{
+			break;
+		}
 
-		// Step-5 Writing to the file specified
-		int temp_write = write(file_fd, output_buffer, strlen(output_buffer));
-		if (temp_write == -1) // generating errors
+		/*Detect '\n' or ASCII value
+			10 in the packet.
+		*/
+		for (i = 0; i < BUFFER_SIZE; i++)
 		{
-			printf("Error: File failed to be written \n");
-			syslog(LOG_ERR, "Error:  File failed to be written= %s. Exiting ", strerror(errno));
-			exit(EXIT_FAILURE);
+
+			if (buff[i] == '\n')
+			{
+				packet_comp = true;
+				i++;
+				printf("data packet receiving completed\n");
+				syslog(LOG_DEBUG, "data packet received");
+				break;
+			}
 		}
-		else if (temp_write != strlen(output_buffer)) // generating errors if partially written
+		j += i; // to get packet size of till null character received
+		data_count += i;
+
+		/*reallocate to a larger buffer now as static buffer can
+			only accomodate upto fixed size*/
+		output_buffer = (char *)realloc(output_buffer, (i + 1));
+		if (output_buffer == NULL)
 		{
-			printf("Error: File not completely written\n");
-			syslog(LOG_ERR, "Error: File not completely written=%s. Exiting ", strerror(errno));
-			exit(EXIT_FAILURE);
+			printf("Realloc failed\n");
+			exit(1);
 		}
 
-		close(file_fd);
+		strncat(output_buffer, buff, i + 1);
 
-		// sending received data to the client back
 		memset(buff, 0, BUFFER_SIZE);
-		file_fd = open(file_data, O_RDONLY);
-		if (file_fd == -1) // generating errors
-		{
-			printf("Error: Opening file\n");
-			syslog(LOG_ERR, "Error: Opening file %s. Exiting ", strerror(errno));
-			exit(EXIT_FAILURE);
-		}
+	}
+
+	// Step-6 Write the data received from client to the server
 
-		// Step-6 Reading from the file
-		send_buffer = (char *)malloc(sizeof(char) * (data_count + 1));
-		memset(send_buffer, 0, data_count + 1);
+	int file_fd = open(file_data, O_APPEND | O_WRONLY);
+	if (file_fd == -1)
+	{
+		printf("File open error for appending\n");
+		exit(1);
+	}
+	ret = pthread_mutex_lock(params->mutex);
+	if (ret)
+	{
+		printf("Mutex lock error before write\n");
+		exit(1);
+	}
 
-		int temp_read = read(file_fd, send_buffer, data_count + 1);
+	int writeret = write(file_fd, output_buffer, strlen(output_buffer));
+	if (writeret == -1)
+	{
+		printf("Error write\n");
+		exit(1);
```

Comment:
@ayswariya088 you probably don't want to kill the entire process when there's an error in your thread.  You are missing cleanup in all of these exit() paths.